### PR TITLE
[WIP] Schema Grants

### PR DESCRIPTION
**Pull request by saraswatpuneet**
# Goal
The goal of this PR is to introduce concept of schema grants which are kind of acl per schema for fine grained controls especially needed in stateful storage operations

With @aramikm 

Closes 

# Discussion

# Checklist
- [ ] Chain spec updated
- [ ] Custom RPC OR Runtime API added/changed? Updated js/api-augment.
- [ ] Design doc(s) updated
- [ ] Tests added
- [ ] Benchmarks added
- [ ] Weights updated

### CODE REVIEW

File: runtime/common/src/constants.rs [hunk 1]
```diff
@@ -132,6 +132,8 @@ parameter_types! {
 pub type SchemasMinModelSizeBytes = ConstU32<8>;
 /// The maximum length of a schema model (in bytes)
 pub type SchemasMaxBytesBoundedVecLimit = ConstU32<65_500>;
+/// The maximum number of grants allowed per schema
+pub type MaxSchemaGrantsPerSchema = ConstU16<1>;
```
Comment (saraswatpuneet) [inline runtime/common/src/constants.rs:136]: just one for now? either AppendOnly or SignatureRequired grants

File: common/primitives/src/schema.rs [hunk 2]
```diff
@@ -35,6 +38,28 @@ pub enum PayloadLocation {
...
+	AppendOnly,
+	/// The metadata of this collection can be modified.
+	SignatureRequired,
+}
+
+impl Default for Grant {
+	fn default() -> Self {
+		Self::AppendOnly
+	}
+}
```
Comment (saraswatpuneet) [inline common/primitives/src/schema.rs:57]: Mostly for mocks/tests gets easier to read

File: runtime/common/src/constants.rs (hunk 1 above)
Comment (saraswatpuneet) [reply runtime/common/src/constants.rs:136]: Or we allow more ?

File: common/primitives/src/macros.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+					.type_params(vec![TypeParameter::new("T", Some(meta_type::<$bitflag_enum>()))])
+					.composite(
+						Fields::unnamed()
+							.field(|f| f.ty::<$size>().type_name(stringify!($bitflag_enum))),
+					)
+			}
+		}
+	};
+}
+// pub(crate) use impl_codec_bitflags;
```
Comment (saraswatpuneet) [inline common/primitives/src/macros.rs:76]: $wrapper wraps a type $size that is used to store a bit flags enum $bitflag_enum and is used to generated derived types with basic features like encoding, decoding etc.

Used for bitflags (grants)

File: common/primitives/src/schema.rs [hunk 4]
```diff
@@ -35,6 +38,28 @@ pub enum PayloadLocation {
...
+/// Support for up to 16 user-enabled features on a collection.
+#[bitflags]
+#[repr(u16)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Copy, Clone, RuntimeDebug, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo)]
+pub enum Grant {
+	/// Items in this collection are transferable.
+	AppendOnly,
+	/// The metadata of this collection can be modified.
+	SignatureRequired,
```
Comment (JoeCap08055) [inline common/primitives/src/schema.rs:50]: How is this going to work:
1. Schema with this property will ALWAYS require a signed encapsulated payload, or
2. Schema with this property may be called normally by owner or delegate, but ALSO may be used with a "one-time delegation" via a signed payload

If (2), then perhaps a different name, like `AllowSignedDelegation`, or something?

Comment (JoeCap08055) [reply common/primitives/src/schema.rs:50]: Also, the documentation comments need to be updated.

File: common/primitives/src/schema.rs (hunk 2 above)
Comment (JoeCap08055) [reply common/primitives/src/schema.rs:57]: This seems dangerous... maybe qualify inside `#[cfg="test"]` or something?

File: pallets/schemas/src/lib.rs [hunk 5]
```diff
@@ -278,9 +285,14 @@ pub mod pallet {
 			model: BoundedVec<u8, T::SchemaModelMaxBytesBoundedVecLimit>,
 			model_type: ModelType,
 			payload_location: PayloadLocation,
+			grants: BoundedVec<Grant, T::SchemaModelMaxBytesBoundedVecLimit>,
```
Comment (saraswatpuneet) [inline pallets/schemas/src/lib.rs:288]: Question: Do we really need a new extrinsic, since grants are optional, this Option should prevent any breaking changes if anyone is actually consuming this ?

File: common/primitives/src/schema.rs (hunk 4 above)
Comment (saraswatpuneet) [reply common/primitives/src/schema.rs:50]: That will be our next story, the focus of this PR is just to update schemas pallet

**Comment by saraswatpuneet** [comment]
@JoeCap08055  when we initially worked on schemas, we said permission are nothing but schemas delegation (publishing rights) ,  Grants we discussed to be called to schema level controls.
We can bring it up in parking lot tomorrow :)

File: common/primitives/src/schema.rs [hunk 6]
```diff
@@ -35,6 +38,28 @@ pub enum PayloadLocation {
 	Paginated,
 }
 
+/// Support for up to 16 user-enabled features on a collection.
+#[bitflags]
+#[repr(u16)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Copy, Clone, RuntimeDebug, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo)]
+pub enum Grant {
```
Comment (aramikm) [inline common/primitives/src/schema.rs]: I agree with Joe to Not call these grants since these are directly associated with Schemas and they are more Properties or Attributes of that schema.

File: common/primitives/src/schema.rs [hunk 7]
```diff
@@ -35,6 +38,28 @@ pub enum PayloadLocation {
...
+	/// Items in this collection are transferable.
+	AppendOnly,
+	/// The metadata of this collection can be modified.
+	SignatureRequired,
+}
+
+impl Default for Grant {
+	fn default() -> Self {
+		Self::AppendOnly
```
Comment (aramikm) [inline common/primitives/src/schema.rs]: I think AppendOnly as the default is a bit odd. Ideally the default should be the default of underlying type `u16` which is 0

File: pallets/schemas/src/lib.rs (hunk 5 above)
Comment (aramikm) [reply pallets/schemas/src/lib.rs]: I think for backwards compatibility we will need a new extrinsic since the signature of the extrinsic in metadata will change

File: pallets/schemas/src/types.rs [hunk 8]
```diff
@@ -18,4 +17,6 @@ where
 	pub model: BoundedVec<u8, MaxModelSize>,
 	/// The payload location
 	pub payload_location: PayloadLocation,
+	/// grants for the schema
+	pub grants: Grants,
```
Comment (aramikm) [inline pallets/schemas/src/types.rs]: I think we will need to create a new Version of Schema to be able to migrate between old version to new.

**Review by aramikm — COMMENTED** [review]
The main logic seems solid just a couple of comments and concerns:

- The Backwards compatibility issue (We probably need new extrinsic) 
- Migration of old Schemas to new ones (we will need this due to rococo and it will allow us to  experience a data migration for the first time)
- `Grants` name doesn't seem very suitable here.

File: pallets/schemas/src/types.rs (hunk 8 above)
Comment (saraswatpuneet) [reply pallets/schemas/src/types.rs]: Understood! yeah did not cross my mind we are deep in it now

**Comment by saraswatpuneet** [comment]
I renamed Grant->SchemaSetting and Grants-> SchemaSettings to keep it simple

File: common/primitives/src/schema.rs (hunk 4 above)
Comment (saraswatpuneet) [reply common/primitives/src/schema.rs:50]: For naming

Should we be explicit about who and what is being signed?

File: pallets/schemas/src/types.rs (hunk 8 above)
Comment (aramikm) [reply pallets/schemas/src/types.rs]: @JoeCap08055 we have to look into what is the preferred way in substrate 
here is an example
https://github.com/paritytech/substrate/blob/master/frame/membership/src/migrations/v4.rs

File: integration-tests/scaffolding/extrinsicHelpers.ts [hunk 9]
```diff
@@ -179,11 +179,16 @@ export class ExtrinsicHelper {
...
     }
 
-    /** Schema Extrinsics */
+    /** Generic Schema Extrinsics */
     public static createSchema(keys: KeyringPair, model: any, modelType: "AvroBinary" | "Parquet", payloadLocation: "OnChain" | "IPFS"| "Itemized" | "Paginated"): Extrinsic {
         return new Extrinsic(() => ExtrinsicHelper.api.tx.schemas.createSchema(JSON.stringify(model), modelType, payloadLocation), keys, ExtrinsicHelper.api.events.schemas.SchemaCreated);
     }
 
+    /** Generic Schema Extrinsics */
+    public static createSchemaWithGrants(keys: KeyringPair, model: any, modelType: "AvroBinary" | "Parquet", payloadLocation: "OnChain" | "IPFS"| "Itemized" | "Paginated", grant: "AppendOnly"| "SignatureRequired"): Extrinsic {
```
Comment (JoeCap08055) [inline integration-tests/scaffolding/extrinsicHelpers.ts:188]: nit: change name to `createSchemaWithSettings`

File: pallets/schemas/src/migrations.rs [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+//! - Rename SchemaV2 to Schema and respective storage once migration is complete
+use super::*;
+use crate::types::{Schema, SchemaV2};
+use codec::{Decode, Encode};
+use common_primitives::schema::{SchemaId, SchemaSettings};
+use frame_support::pallet_prelude::Weight;
+
+pub fn migrate_schema_to_schema_v2<T: Config>() -> Weight {
+	let mut weight: Weight = Weight::zero();
+	let schema_count = <Schemas<T>>::iter().count();
```
Comment (JoeCap08055) [inline pallets/schemas/src/migrations.rs:15]: Maybe more efficient to use `iter().enumerate()` here instead of `count()`? Would save having to re-fetch the schema later on.

File: pallets/schemas/src/lib.rs [hunk 11]
```diff
@@ -174,6 +181,19 @@ pub mod pallet {
 		OptionQuery,
 	>;
 
+	/// Storage for SchemaV2
+	/// - Key: Schema Id
+	/// - Value: [`SchemaV2`](SchemaV2)
+	#[pallet::storage]
+	#[pallet::getter(fn get_schema_v2)]
+	pub(super) type SchemasV2<T: Config> = StorageMap<
```
Comment (JoeCap08055) [inline pallets/schemas/src/lib.rs:189]: Just a thought: If we rename the current/original `Schemas` storage to `SchemasV1` and call the new one `Schemas`, it will ensure that any existing references will use the new storage (and eliminate the need for post-migration renaming/code modification). (Will still need to delete the V1 storage post-migration, but that's all).

File: pallets/schemas/src/types.rs (hunk 8 above)
Comment (saraswatpuneet) [reply pallets/schemas/src/types.rs]: There are few different kinds of migrations, like runtime etc. since we are simply translating our schema to new type it will just be a simple hook 

Example for such https://github.com/paritytech/polkadot/pull/4545

File: pallets/schemas/src/lib.rs [hunk 12]
```diff
@@ -263,6 +283,41 @@ pub mod pallet {
...
+		/// * [`Error::InvalidSchema`] - Schema is malformed in some way
+		/// * [`Error::SchemaCountOverflow`] - The schema count has exceeded its bounds
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::create_schema(model.len() as u32 + settings.len() as u32))]
+		pub fn create_schema_with_settings(
+			origin: OriginFor<T>,
+			model: BoundedVec<u8, T::SchemaModelMaxBytesBoundedVecLimit>,
+			model_type: ModelType,
+			payload_location: PayloadLocation,
+			settings: BoundedVec<SchemaSetting, T::MaxSchemaSettingsPerSchema>,
```
[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply pallets/schemas/src/lib.rs:297]: That would break the metadata so we would rather create a new extrinsic then forcing downstream to update metadata if they arent using stateful related stuff when we upgrade the chain and also we have a clear visibility.

open to suggestion

File: pallets/schemas/src/migrations.rs [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+		weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+		Some(Schema {
+			model_type: old_schema.model_type,
+			model: old_schema.model,
+			payload_location: old_schema.payload_location,
+			settings: SchemaSettings::all_disabled(),
+		})
+	});
+	weight
+}
```
Comment (saraswatpuneet) [inline pallets/schemas/src/migrations.rs:23]: @aramikm  so since we are simply translating values, this is some places where parity simply uses on_runtime_upgrade hook to run migrations

File: pallets/schemas/src/lib.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply pallets/schemas/src/lib.rs:297]: And keeping option then it just behaves like original one, keeping it simple

File: pallets/schemas/src/lib.rs [hunk 14]
```diff
@@ -193,6 +200,13 @@ pub mod pallet {
 		}
 	}
 
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_runtime_upgrade() -> Weight {
+			migrations::migrate_schemas_with_additional_settings::<T>()
+		}
+	}
+
```
Comment (saraswatpuneet) [inline pallets/schemas/src/lib.rs:209]: This i havent not tested yet, but thats the plan, to do some upgrades and check if migration works, then we can add some try-runtime flags to have a test case

File: pallets/schemas/src/migrations.rs [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+//! Note: Post migration, this file should be deleted.
+use crate::{types::Schema, *};
+use common_primitives::schema::SchemaSettings;
+use frame_support::pallet_prelude::Weight;
+
+pub fn migrate_schemas_with_additional_settings<T: Config>() -> Weight {
+	let mut weight: Weight = Weight::zero();
+	//translate_values of Schemas storage add settings: SchemaSettings::all_disabled()
+	//to all schemas
+	<Schemas<T>>::translate_values(|old_schema: Schema<T::SchemaModelMaxBytesBoundedVecLimit>| {
```
Comment (aramikm) [inline pallets/schemas/src/migrations.rs:13]: I don't see us using different Schema types for old and new ones. How does substrate know how to decode old values? I'm not sure if this will work

Comment (aramikm) [reply pallets/schemas/src/migrations.rs:13]: I also like enwrapping the migration inside a mod version 

```rust
mod v1{
    // migration logic
}
```
example https://github.com/paritytech/substrate/blob/master/frame/contracts/src/migration.rs#L87

Comment (saraswatpuneet) [reply pallets/schemas/src/migrations.rs:13]: In general through any language or db i expect decoders to just populate fields that exists, hoping its the same here as I expect this to working similar to going through 
https://github.com/paritytech/polkadot/pull/4545

Comment (saraswatpuneet) [reply pallets/schemas/src/migrations.rs:13]: I also like the ides of modulating it, will do after these arduous upgrade testing

Comment (aramikm) [reply pallets/schemas/src/migrations.rs:13]: but even in that example there is 2 versions for **SessionInfo**
```

<Pallet<T> as Store>::Sessions::translate_values(|old: primitives::**v1**::SessionInfo| {
		vs += 1;
		Some(primitives::**v2**::SessionInfo::from(old))
```

**Comment by saraswatpuneet** [comment]
A full upgrade test to make sure we have successful migration

https://docs.google.com/document/d/1_CrTeVKh_sq_twsB60WNMd1SKnyWxOL_AW6I9a-5XKI/edit?usp=sharing

**Comment by JoeCap08055** [comment]
> A full upgrade test to make sure we have successful migration
> 
> https://docs.google.com/document/d/1_CrTeVKh_sq_twsB60WNMd1SKnyWxOL_AW6I9a-5XKI/edit?usp=sharing

As part of the validation, should verify that all schemas were upgraded... the document only shows 1 schema queried after upgrade (maybe just not shown in the doc)

File: common/primitives/src/schema.rs [hunk 16]
```diff
@@ -48,6 +67,8 @@ pub struct SchemaResponse {
 	pub model_type: ModelType,
 	/// The payload location
 	pub payload_location: PayloadLocation,
+	/// grants for the schema
+	pub settings: Vec<SchemaSetting>,
```
[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply common/primitives/src/schema.rs:71]: @aramikm  i am a bit divided here, to support SchemaSetting directly we would definitely need a custom serializer and de for Bitflags<SchemaSetting, u16> given that what essentially SchemaSettings is

I was under the impression to return actual types as vec to prevent doing all above 

What you suggest, or i can set a ticket out to fix this?

[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply common/primitives/src/schema.rs:71]: Another thing is in our users mind they are setting some values and exactly what they get back I think that will keep consistency than serializing flags what you think?

[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply common/primitives/src/schema.rs:71]: Looks like this never translates back I get nil so this needs fix before merge 

Will do that

[the remark this answers is not in the document]
Comment (aramikm) [reply common/primitives/src/schema.rs:73]: I think this is fine. It might be confusing if the input and output is not consistent from user perspective as you pointed out.

File: pallets/schemas/src/lib.rs [hunk 17]
```diff
@@ -193,6 +200,13 @@ pub mod pallet {
 		}
 	}
 
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_runtime_upgrade() -> Weight {
+			migrations::v1::migrate_schemas_with_additional_settings::<T>()
```
Comment (aramikm) [inline pallets/schemas/src/lib.rs:206]: it does seem like we will need a cleanup ticket to remove all these runtime upgrade logic

File: pallets/schemas/src/migrations.rs (hunk 13 above)
Comment (aramikm) [reply pallets/schemas/src/migrations.rs:43]: Is there any possibly that a runtime migration hook might get triggered more than 1 time?

**Review by aramikm — APPROVED** [review]
Looks great to me. Let's make sure somebody else with migration experience also checks this before merge.

File: pallets/schemas/src/lib.rs (hunk 17 above)
Comment (saraswatpuneet) [reply pallets/schemas/src/lib.rs:206]: yep https://github.com/LibertyDSNP/frequency/issues/1106

**Review by JoeCap08055 — APPROVED** [review]

File: pallets/schemas/src/migrations.rs (hunk 13 above)
Comment (shannonwells) [reply pallets/schemas/src/migrations.rs:43]: For safety reasons, a migration hook should be designed to be idempotent even if you expect to remove it before the next runtime upgrade.  There should be some check like for if it's OldSchema or not before calling `translate_values`.

File: pallets/schemas/src/lib.rs (hunk 14 above)
Comment (shannonwells) [reply pallets/schemas/src/lib.rs:209]: try-runtime is meant to be used for testing, esp. storage migrations.  If you have a look at the [remove sudo storage PR](https://github.com/LibertyDSNP/frequency/pull/1066) it has a bunch of logging in the try-runtime, but logging is a good idea anyway in case something goes wrong.

File: pallets/schemas/src/tests.rs [hunk 18]
```diff
@@ -350,3 +350,27 @@ fn dsnp_broadcast() {
 	);
 	assert_ok!(result);
 }
+
+#[test]
+fn create_schema_with_settings_should_work() {
```
Comment (shannonwells) [inline pallets/schemas/src/tests.rs:355]: Error tests?

File: pallets/schemas/src/lib.rs [hunk 19]
```diff
@@ -335,6 +386,23 @@ pub mod pallet {
 
 			Ok(next)
 		}
+
+		fn validate_schema_model(
```
Comment (shannonwells) [inline pallets/schemas/src/lib.rs:390]: 🏅

**Review by shannonwells — COMMENTED** [review]
Please see comments

File: pallets/schemas/src/migrations.rs (hunk 13 above)
Comment (saraswatpuneet) [reply pallets/schemas/src/migrations.rs:43]: @shannonwells I believe this `|old_schema: OldSchema<T::SchemaModelMaxBytesBoundedVecLimit>|` cast will prevent translation of any schemas that dont fit the OldSchema model since any new schemas wont decode to OldSchema type

Or you mean its safe to` iter -> check -> update?`

File: pallets/schemas/src/tests.rs (hunk 18 above)
Comment (saraswatpuneet) [reply pallets/schemas/src/tests.rs:355]: Since schema dont directly check settings but consumers of schema does, I have added more tests and negative tests for SchemaSetting in stateful pallet

File: pallets/stateful-storage/src/lib.rs [hunk 20]
```diff
@@ -430,55 +428,96 @@ impl<T: Config> Pallet<T> {
...
+		// Ensure that the schema allows signed payloads.
+		// If so, calling extrinsic must be of signature type.
+		if schema.settings.contains(&SchemaSetting::SignatureRequired) {
+			ensure!(is_payload_signed, Error::<T>::SchemaNotSupported);
+		}
+
+		// Ensure that the schema does not allow deletion for AppendOnly SchemaSetting.
+		if schema.settings.contains(&SchemaSetting::AppendOnly) {
+			ensure!(!is_deleting, Error::<T>::SchemaNotSupported);
+		}
```
Comment (saraswatpuneet) [inline pallets/stateful-storage/src/lib.rs:508]: @aramikm  what you think, wanted to keep it simple, given we will circle back at validation

File: runtime/frequency/src/lib.rs [hunk 21]
```diff
@@ -1100,6 +1115,38 @@ cumulus_pallet_parachain_system::register_validate_block! {
...
+	fn post_upgrade(_state: Vec<u8>) -> Result<(), &'static str> {
+		let weight = pallet_schemas::migrations::v1::post_migrate_schemas_to_v1::<Runtime>();
+		log::info!("post_upgrade weight: {:?}", weight);
+		Ok(())
+	}
+}
+// ==============================================
+//       END RUNTIME STORAGE MIGRATION: Schemas
+// ==============================================
+
```
Comment (saraswatpuneet) [inline runtime/frequency/src/lib.rs:1149]: @shannonwells  tested with this and works fine, i see schemas migrated

File: pallets/schemas/src/migrations.rs (hunk 13 above)
Comment (aramikm) [reply pallets/schemas/src/migrations.rs:64]: I think she means checking even before iteration. One thing I've seen some pallets do it that they have a separate storage to keep track of current version of migration and update it after the migration is complete. So if the process runs again it would check with that version and not run if it's the latest version. There might be other ways also to achieve this.

File: pallets/stateful-storage/src/benchmarking.rs [hunk 22]
```diff
@@ -11,7 +11,7 @@ use frame_system::RawOrigin;
 use sp_core::bounded::BoundedVec;
 use stateful_child_tree::StatefulChildTree;
 
-pub const ITEMIZED_SCHEMA: SchemaId = 100; // keep in sync with mock.rs. TODO: refactor
+pub const ITEMIZED_APPEND_ONLY_SCHEMA: SchemaId = 100; // keep in sync with mock.rs. TODO: refactor
```
Comment (aramikm) [inline pallets/stateful-storage/src/benchmarking.rs:14]: these are refactored. Please make sure to use the `test_common.rs` in the pallet.

File: pallets/schemas/src/migrations.rs (hunk 13 above)
Comment (saraswatpuneet) [reply pallets/schemas/src/migrations.rs:64]: I see, yes I remember seeing that around, will redo this part

File: pallets/schemas/src/lib.rs [hunk 23]
```diff
@@ -143,6 +150,7 @@ pub mod pallet {
 
 	#[pallet::pallet]
 	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(SCHEMA_STORAGE_VERSION)]
```
Comment (saraswatpuneet) [inline pallets/schemas/src/lib.rs:153]: Storage version will be checked at the time of migration and anything older will be translated while newer ignored

Comment (aramikm) [reply pallets/schemas/src/lib.rs:153]: I'm curious how this works under the hood. Just want to make sure there are no surprises

**Review by aramikm — APPROVED** [review]
Looks good! Ship it ✅
